synedrion/src/paillier/encryption.rs

```diff
@@ -207,33 +253,42 @@ impl<P: PaillierParams> Ciphertext<P> {
         // `C^phi mod N^2` may be 0 if `C == N`, which is very unlikely for large `N`.
         // Note that `C^phi mod N^2 / N < N`, so we can unwrap when converting to `Uint`
         // (because `N` itself fits into `Uint`).
-        let x = P::Uint::try_from_wide(
-            (self.ciphertext.pow_bounded(totient_wide.expose_secret())
-                - P::WideUintMod::one(pk.monty_params_mod_n_squared().clone()))
-            .retrieve()
-                / pk.modulus_wide_nonzero(),
-        )
-        .expect("the value is within `Uint` limtis by construction");
+
+        // Calculate `C^phi mod N^2`. The result is already secret.
+        // The exponent may leave traces on the stack in the `pow` implementation in `crypto-bigint`
+        // (since it'll probably be decomposed with a small radix), but we can't do much about that.
+        let t = Secret::init_with(|| self.ciphertext.pow_bounded(totient_wide.expose_secret()));
+        let one = P::WideUintMod::one(pk.monty_params_mod_n_squared().clone());
+        let x = (t - &one).retrieve() / pk.modulus_wide_nonzero();
+        let x = Secret::init_with(|| {
+            P::Uint::try_from_wide(x.expose_secret()).expect("the value is within `Uint` limtis by construction")
+        });
 
         let x_mod = x.to_montgomery(pk.monty_params_mod_n());
 
-        (x_mod * sk.inv_totient().expose_secret()).retrieve()
+        (x_mod * sk.inv_totient()).retrieve()
     }
 
     /// Decrypts this ciphertext assuming that the plaintext is in range `[-N/2, N/2)`.
-    pub fn decrypt_signed(&self, sk: &SecretKeyPaillier<P>) -> Signed<P::Uint> {
+    pub fn decrypt_signed(&self, sk: &SecretKeyPaillier<P>) -> Secret<Signed<P::Uint>> {
         assert_eq!(sk.public_key(), &self.pk);
 
         let pk = sk.public_key();
-        let positive_result = self.decrypt(sk); // Note that this is in range `[0, N)`
-        let negative_result = pk.modulus().wrapping_sub(&positive_result);
-        let is_negative = Choice::from((positive_result > pk.modulus().wrapping_shr_vartime(1)) as u8);
 
-        let mut result = Signed::new_from_unsigned(
-            P::Uint::conditional_select(&positive_result, &negative_result, is_negative),
-            P::MODULUS_BITS - 1,
-        )
-        .expect("the value is within `[-2^(MODULUS_BITS-1), 2^(MODULUS_BITS-1)]` by construction");
+        // Note that this is in range `[0, N)`
+        let positive_result = self.decrypt(sk);
+
+        // Can't define a `Sub<Secret>` for `Uint`, so have to re-wrap manually.
```

Comment:
What's the blocker for that?

Reply:
I would be implementing a foreign trait (`Sub`) for a foreign type (or for a possibly foreign type, something like `impl<T: Zeroize + for<'a> SubAssign<&'a T>> Sub<Secret<T>> for T`), which is not allowed.